File: tests/benchmarks/resolution/fixtures/clojure/service.clj
(ns app.service
  (:require [app.repository :as repository]
            [app.validators :as validators]))

(defn create-user [repo name email]
  (let [valid-name (validators/validate-name name)
        valid-email (validators/validate-email email)]
    (if (and valid-name valid-email)
      (repository/save repo name email)
      {:error "validation failed"})))

(defn get-user [repo id]
  (repository/find-by-id repo id))

(defn remove-user [repo id]
  (repository/delete repo id))

(defn summary [repo]
  (let [cnt (repository/count repo)]
    (format-summary cnt)))

(defn- format-summary [cnt]
  (str "repository contains " cnt " users"))

File: tests/benchmarks/resolution/fixtures/haskell/Service.hs
module Service
  ( createUser
  , getUser
  , removeUser
  , summary
  ) where

import qualified Repository as Repo
import Validators (validateEmail, validateName, validateAge)
import Data.Map.Strict (Map)

type UserId = String
type UserRecord = (String, String, Int)
type Store = Map UserId UserRecord

validateUser :: String -> String -> Int -> Bool
validateUser name email age =
  validateName name && validateEmail email && validateAge age

createUser :: UserId -> String -> String -> Int -> Store -> Either String Store
createUser uid name email age store =
  if validateUser name email age
    then Right (Repo.save uid (name, email, age) store)
    else Left "Validation failed"

getUser :: UserId -> Store -> Maybe UserRecord
getUser uid store = Repo.findById uid store

removeUser :: UserId -> Store -> Store
removeUser uid store = Repo.delete uid store

summary :: Store -> Int
summary store = Repo.count store
